File: Data Structures & Algorithms/maximum-product-subarray/submission-5.java
class Solution {
    public int maxProduct(int[] nums) {
        int res = nums[0];
        int mini = 1;
        int maxi = 1;

        for (int elem : nums) {
            int miniClone = Math.min(Math.min(elem, mini * elem), maxi * elem);
            int maxiClone = Math.max(Math.max(elem, mini * elem), maxi * elem);
            res = Math.max(res, maxiClone);
            maxi = maxiClone;
            mini = miniClone;
        }
        return res;

    }
}

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-5.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        std::unordered_map<int, int> map;
        int res = 0;
        for (int x : nums) {
            if (map[x]) continue;
            map[x] = map[x - 1] + map[x + 1] + 1;
            map[x - map[x - 1]] = map[x];
            map[x + map[x + 1]] = map[x];
            res = std::max(res, map[x]);
        }
        return res;
    }
};
